drivers/sdhc/rcar_mmc.c

```diff
@@ -0,0 +1,2388 @@
+		LOG_ERR("%s: block size (%u) must not be bigger than 512 bytes and equal to zero",
+			dev->name, data->block_size);
+		return -EINVAL;
+	}
+
+	cmd_reg = rcar_mmc_read_reg32(dev, RCAR_MMC_CMD);
+	if (cmd_reg & RCAR_MMC_CMD_MULTI) {
+		/* CMD12 is automatically issued at multiple block transfer */
+		if (!(cmd_reg & RCAR_MMC_CMD_NOSTOP) && data->block_size != 512) {
+			LOG_ERR("%s: illegal block size (%u) for multi-block xref with CMD12",
+				dev->name, data->block_size);
+			return -EINVAL;
+		}
+
+		switch (data->block_size) {
+		case 32:
+		case 64:
+		case 128:
+		case 256:
+		case 512:
+			break;
+		default:
+			LOG_ERR("%s: illegal block size (%u) for multi-block xref without CMD12",
+				dev->name, data->block_size);
+			return -EINVAL;
+		}
+	}
+
+	if (data->block_size == 1 && dev_data->host_io.bus_width == SDHC_BUS_WIDTH8BIT) {
+		LOG_ERR("%s: block size can't be equal to 1 with 8-bits bus width", dev->name);
+		return -EINVAL;
+	}
+
+	for (block = 0; block < data->blocks; block++) {
+		uint8_t *buf = (uint8_t *)data->data + (block * data->block_size);
+		uint32_t info2_reg;
+		uint16_t w_off; /* word offset in a block */
+		uint64_t start_block_xref_us = k_ticks_to_us_ceil64(k_uptime_ticks());
+
+		/* wait until the buffer is filled with data */
+		ret = rcar_mmc_poll_reg_flags_check_err(dev, RCAR_MMC_INFO2, info2_poll_flag,
+							info2_poll_flag, true, false,
+							remaining_timeout_us);
+		if (ret) {
+			return ret;
+		}
+
+		/* clear write/read buffer ready flag */
+		info2_reg = rcar_mmc_read_reg32(dev, RCAR_MMC_INFO2);
+		info2_reg &= ~info2_poll_flag;
+		rcar_mmc_write_reg32(dev, RCAR_MMC_INFO2, info2_reg);
+
+		for (w_off = 0; w_off < aligned_block_size; w_off += sd_buf0_size) {
+			uint64_t buf0 = 0ULL;
+			uint8_t copy_size = MIN(sd_buf0_size, data->block_size - w_off);
+
+			if (is_read) {
+				buf0 = rcar_mmc_read_buf0(dev);
+				memcpy(buf + w_off, &buf0, copy_size);
+			} else {
+				memcpy(&buf0, buf + w_off, copy_size);
+				rcar_mmc_write_buf0(dev, buf0);
+			}
+		}
+
+		remaining_timeout_us -=
+			k_ticks_to_us_ceil64(k_uptime_ticks()) - start_block_xref_us;
+		if (remaining_timeout_us < 0) {
+			return -ETIMEDOUT;
+		}
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Transmit/Receive data to/from MMC
+ *
+ * Sends/Receives data to/from the MMC controller.
+ *
+ * @note in/out parameters should be checked by a caller function.
+ *
+ * @param dev MMC device
+ * @param data MMC data buffer for tx/rx
+ * @param is_read it is read or write operation
+ *
+ * @retval 0 tx/rx was successful
+ * @retval -EINVAL: invalid block size
+ * @retval -ETIMEDOUT: timed out while tx/rx
+ * @retval -EIO: I/O error
+ * @retval -EILSEQ: communication out of sync
+ */
+static int rcar_mmc_rx_tx_data(const struct device *dev, struct sdhc_data *data, bool is_read)
+{
+	struct mmc_rcar_data *dev_data = dev->data;
+	uint32_t info1_reg;
+	int ret = 0;
+
+	if (dev_data->dma_support && ((uintptr_t)data->data % CONFIG_SDHC_BUFFER_ALIGNMENT == 0) &&
```

Comment:
The SD subsystem should insure all data buffers passed to the driver are aligned to `CONFIG_SDHC_BUFFER_ALIGNMENT`, see here: https://github.com/zephyrproject-rtos/zephyr/blob/18636af6d6658cf66e5f055901ffe2515116d329/subsys/sd/sd_ops.c

So it should be possible to always use the DMA function when `dev_data->dma_support` is true unless I'm missing something specific to this controller?